src/py/mat3ra/made/tools/utils.py

```diff
@@ -322,3 +326,127 @@ def plane(plane_normal: List[float], plane_point_coordinate: List[float]):
             plane_normal=plane_normal,
             plane_point_coordinate=plane_point_coordinate,
         )
+
+
+class PerturbationFunctionHolder:
+    @staticmethod
+    def get_coord_transformation(perturbation_json: dict) -> Callable:
+        new_perturbation_json = perturbation_json.copy()
+        name = new_perturbation_json.pop("type")
+        name_to_function_map = {"sine_wave": PerturbationFunctionHolder.sine_wave_transform_coordinates}
+        return name_to_function_map[name](**new_perturbation_json)
+
+    @staticmethod
+    def sine_wave(
+        amplitude: float = 0.1,
+        wavelength: float = 1,
+        phase: float = 0,
+        axis: Optional[Literal["x", "y"]] = "x",
+    ) -> Tuple[Callable[[List[float]], List[float]], Dict]:
+        """
+        Deform a coordinate using a sine wave.
+        Args:
+            amplitude (float): The amplitude of the sine wave in cartesian coordinates.
+            wavelength (float): The wavelength of the sine wave in cartesian coordinates.
+            phase (float): The phase of the sine wave in cartesian coordinates.
+            axis (str): The axis of the direction of the sine wave.
+
+        Returns:
+            Tuple[Callable[[List[float]], List[float]], Dict]: The perturbation function and its configuration
+        """
+        if axis in AXIS_TO_INDEX_MAP:
+            index = AXIS_TO_INDEX_MAP[axis]
+
+        def deformation(coordinate: List[float]):
+            return [
+                coordinate[0],
+                coordinate[1],
+                coordinate[2] + amplitude * np.sin(2 * np.pi * coordinate[index] / wavelength + phase),
+            ]
+
+        config = {"type": "sine_wave", "amplitude": amplitude, "wavelength": wavelength, "phase": phase, "axis": axis}
+
+        return deformation, config
+
+    @staticmethod
+    def sine_wave_transform_coordinates(
+        amplitude: float, wavelength: float, phase: float, axis: Literal["x", "y"]
+    ) -> Callable[[List[float]], List[float]]:
+        """
+        Transform coordinates to preserve the distance between points on a sine wave.
+        Achieved by calculating the integral of the length between [0,0,0] and given coordinate.
+
+        Args:
+            amplitude (float): The amplitude of the sine wave in cartesian coordinates.
+            wavelength (float): The wavelength of the sine wave in cartesian coordinates.
+            phase (float): The phase of the sine wave in cartesian coordinates.
+            axis (str): The axis of the direction of the sine wave.
+
+        Returns:
+            Callable[[List[float]], List[float]]: The coordinates transformation function.
+        """
+
+        def sine_wave_diff(w: float, amplitude: float, wavelength: float, phase: float) -> float:
```

Comment:
We need to have a generic approach to this where we define a function and its differential plus length integral either with SymPy or explicitly